README.md
SockJS-tornado server
=====================

SockJS-tornado is a Python server side counterpart of [SockJS-client browser library](https://github.com/sockjs/sockjs-client)
running on top of [Tornado](http://tornadoweb.org>) framework.

Simplified echo SockJS server could look more or less like::

    from tornado import web, ioloop
    from sockjs.tornado import SockJSRouter, SockJSConnection

    class EchoConnection(SockJSConnection):
        def on_message(self, msg):
            self.send(msg)

    if __name__ == '__main__':
        EchoRouter = SockJSRouter(EchoConnection, '/echo')

        app = web.Application(EchoRouter.urls)
        app.listen(9999)
        ioloop.IOLoop.instance().start()

(Take look at [examples](https://github.com/MrJoes/sockjs-tornado/tree/master/examples>) for a complete version).

Subscribe to [SockJS mailing list](https://groups.google.com/forum/#!forum/sockjs>) for discussions and support.

SockJS-tornado API
------------------

SockJS provides slightly different API than _tornado.websocket_. Main differences are:

1.  Depending on transport, actual client connection might or might not be there. So, there is no _self.request_ and
    other _tornado.web.RequestHandler_ properties.
2.  Changed _open_ callback name to _on\_open_ to be more consistent with other callbacks.
3.  Instead of _write\_message_, all messages are sent using _send_ method. Just in case, _send_ in _tornado.web.RequestHandler_
    sends raw data over the connection, without encoding it.
4.  There is handy _broadcast_ function, which accepts list (or iterator) of clients and message to send.

Settings
--------

You can pass various settings to the _SockJSRouter_, in a dictionary::

    MyRouter = SockJSRouter(MyConnection, '/my', dict(disabled_transports=['websocket']))

Following options are supported:

<table>
    <tr>
        <td>*Name*</td>
        <td>*Description*</td>
        <td>*Default Value*</td>
    </tr>
    <tr>
        <td>session_check_interval</td>
        <td>How often sessions are verified, in seconds. Lower the value, more resources sockjs-tornado will use to drop expired sessions, but will be more precise.</td>
        <td>1</td>
    </tr>
    <tr>
        <td>disconnect_delay</td>
        <td>How long does it takes to consider session dead, in seconds.</td>
        <td>5</td>
    </tr>
    <tr>
        <td>heartbeat_delay</td>
        <td>How often sockjs-tornado should send heartbeats to the client, in seconds.</td>
        <td>25</td>
    </tr>
    <tr>
        <td>disabled_transports</td>
        <td>List of disabled transports. Possible values are: 'websocket', 'xhr', 'xhr_streaming', 'jsonp', 'eventsource' and 'htmlfile'</td>
        <td>[]</td>
    </tr>
    <tr>
        <td>sockjs_url</td>
        <td>Location of the sockjs client script. By default points to the SockJS CDN.</td>
        <td>http://cdn.sockjs.org/sockjs-0.1.2.min.js</td>
    </tr>
</table>


Deployment
----------

sockjs-tornado properly works behind haproxy and it is recommended deployment approach.

Sample configuration file can be found [here](https://raw.github.com/sockjs/sockjs-node/master/examples/haproxy.cfg).

If your log is full of "WARNING: Connection closed by the client", pass _no\_keep\_alive_ as _True_ to _HTTPServer_ constructor::

    HTTPServer(app, no_keep_alive=True).listen(port)

or::

    app.listen(port, no_keep_alive=True)



doc/changelog.rst
Changelog
---------

1.0.7
-----

 - Tornado 6.x compatibility fix. Minimum supported Tornado version is 4.x now.
 - XSS fix

1.0.6
-----

 - XSS security fix for the HTMLFILE transport

1.0.2
~~~~~
 - Tornado 5.x compatibility fix

1.0.1
~~~~~
 - Tornado 4.x compatibility fix

1.0.0
~~~~~
 - Major version bump. sockjs-tornado can be considered mature
 - Python 3 support
 - Tornado 3.0 support on both Python 2 and Python 3
 - Bug fixes


0.0.5
~~~~~
 - iOS 6 support
 - SockJS 0.3.3 client support
 - Some headers and request URI path is now passed to on_open handler
 - Added binary transport support for raw websocket transport
 - Minor multiplex sample fixes

0.0.4
~~~~~

 - Added support for older simplejson library versions
 - Fixed installation script

0.0.3
~~~~~

 - SockJS 0.3 support
 - Connection multiplexing sample
 - IP validation is now optional

0.0.2
~~~~~

 - SockJS 0.2 support and compatibility fixes

0.0.1
~~~~~

Initial release.


sockjs/tornado/router.py
# -*- coding: utf-8 -*-
"""
    sockjs.tornado.router
    ~~~~~~~~~~~~~~~~~~~~~

    SockJS protocol router implementation.
"""

from tornado import ioloop, version_info

from sockjs.tornado import (
    transports,
    session,
    sessioncontainer,
    static,
    stats,
    proto
)


DEFAULT_SETTINGS = {
    # Sessions check interval in seconds
    'session_check_interval': 1,
    # Session expiration in seconds
    'disconnect_delay': 5,
    # Heartbeat time in seconds. Do not change this value unless
    # you absolutely sure that new value will work.
    'heartbeat_delay': 25,
    # Websocket heartbeat check interval
    'heartbeat_check_delay': 10,
    # Enabled protocols
    'disabled_transports': [],
    # SockJS location
    'sockjs_url': 'https://cdn.jsdelivr.net/sockjs/0.3/sockjs.min.js',
    # Max response body size
    'response_limit': 128 * 1024,
    # Enable or disable JSESSIONID cookie handling
    'jsessionid': True,
    # Should sockjs-tornado flush messages immediately or queue then and
    # flush on next ioloop tick
    'immediate_flush': True,
    # Enable or disable Nagle for persistent transports
    'disable_nagle': True,
    # Enable IP checks for polling transports. If enabled, all subsequent
    # polling calls should be from the same IP address.
    'verify_ip': True,
    # list of allowed origins for websocket connections
    # or "*" - accept all websocket connections
    'websocket_allow_origin': "*"
}

GLOBAL_HANDLERS = [
    ('xhr_send', transports.XhrSendHandler),
    ('jsonp_send', transports.JSONPSendHandler)
]

TRANSPORTS = {
    'websocket': transports.WebSocketTransport,
    'xhr': transports.XhrPollingTransport,
    'xhr_streaming': transports.XhrStreamingTransport,
    'jsonp': transports.JSONPTransport,
    'eventsource': transports.EventSourceTransport,
    'htmlfile': transports.HtmlFileTransport
}

STATIC_HANDLERS = {
    '/chunking_test': static.ChunkingTestHandler,
    '/info': static.InfoHandler,
    '/iframe[0-9-.a-z_]*.html': static.IFrameHandler,
    '/websocket': transports.RawWebSocketTransport,
    '/?': static.GreetingsHandler
}


class SockJSRouter(object):
    """SockJS protocol router"""
    def __init__(self,
                 connection,
                 prefix='',
                 user_settings=dict(),
                 io_loop=None,
                 session_kls=None):
        """Constructor.

        `connection`
            SockJSConnection class
        `prefix`
            Connection prefix
        `user_settings`
            Settings dictionary
        `io_loop`
            Optional IOLoop instance
        """

        # TODO: Version check
        if version_info[0] < 2:
            raise Exception('sockjs-tornado requires Tornado 2.0 or higher.')

        # Store connection class
        self._connection = connection

        # Initialize io_loop
        self.io_loop = io_loop or ioloop.IOLoop.instance()

        # Settings
        self.settings = DEFAULT_SETTINGS.copy()
        if user_settings:
            self.settings.update(user_settings)

        self.websockets_enabled = 'websocket' not in self.settings['disabled_transports']
        self.cookie_needed = self.settings['jsessionid']

        # Sessions
        self._session_kls = session_kls if session_kls else session.Session
        self._sessions = sessioncontainer.SessionContainer()

        check_interval = self.settings['session_check_interval'] * 1000
        self._sessions_cleanup = ioloop.PeriodicCallback(self._sessions.expire,
                                                         check_interval,
                                                         self.io_loop)
        self._sessions_cleanup.start()

        # Stats
        self.stats = stats.StatsCollector(self.io_loop)

        # Initialize URLs
        base = prefix + r'/[^/.]+/(?P<session_id>[^/.]+)'

        # Generate global handler URLs
        self._transport_urls = [('%s/%s$' % (base, p[0]), p[1], dict(server=self))
                                for p in GLOBAL_HANDLERS]

        for k, v in TRANSPORTS.items():
            if k in self.settings['disabled_transports']:
                continue

            # Only version 1 is supported
            self._transport_urls.append(
                (r'%s/%s$' % (base, k),
                 v,
                 dict(server=self))
            )

        # Generate static URLs
        self._transport_urls.extend([('%s%s' % (prefix, k), v, dict(server=self))
                                     for k, v in STATIC_HANDLERS.items()])

    @property
    def urls(self):
        """List of the URLs to be added to the Tornado application"""
        return self._transport_urls

    def apply_routes(self, routes):
        """Feed list of the URLs to the routes list. Returns list"""
        routes.extend(self._transport_urls)
        return routes

    def create_session(self, session_id, register=True):
        """Creates new session object and returns it.

        `request`
            Request that created the session. Will be used to get query string
            parameters and cookies
        `register`
            Should be session registered in a storage. Websockets don't
            need it.
        """
        # TODO: Possible optimization here for settings.get
        s = self._session_kls(self._connection,
                            self,
                            session_id,
                            self.settings.get('disconnect_delay')
                            )

        if register:
            self._sessions.add(s)

        return s

    def get_session(self, session_id):
        """Get session by session id

        `session_id`
            Session id
        """
        return self._sessions.get(session_id)

    def get_connection_class(self):
        """Return associated connection class"""
        return self._connection

    # Broadcast helper
    def broadcast(self, clients, msg):
        """Optimized `broadcast` implementation. Depending on type of the session, will json-encode
        message once and will call either `send_message` or `send_jsonifed`.

        `clients`
            Clients iterable
        `msg`
            Message to send
        """
        json_msg = None

        count = 0

        for c in clients:
            sess = c.session
            if not sess.is_closed:
                if sess.send_expects_json:
                    if json_msg is None:
                        json_msg = proto.json_encode(msg)
                    sess.send_jsonified(json_msg, False)
                else:
                    sess.send_message(msg, stats=False)

                count += 1

        self.stats.on_pack_sent(count)
